stream/src/main/scala/org/apache/pekko/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -614,14 +607,31 @@ import pekko.util.ccompat.JavaConverters._
 
       override def onDownstreamFinish(cause: Throwable): Unit = {
         substreamCancelled = true
-        if (isClosed(in) || propagateSubstreamCancel(cause)) {
-          cancelStage(cause)
-        } else {
-          // Start draining
-          if (!hasBeenPulled(in)) pull(in)
+        decider(cause) match {
+          case Supervision.Stop =>
+            cancelStage(cause)
+          case Supervision.Resume =>
+            if (isClosed(in)) cancelStage(cause)
+            else {
+              // Start draining
+              if (!hasBeenPulled(in)) pull(in)
+            }
+          case Supervision.Restart =>
+            if (isClosed(in)) completeStage()
+            else {
+              restartState()
+              // Start draining
+              if (!hasBeenPulled(in)) pull(in)
+            }
```

Comment:
I'm not sure about this, it was restart the `sink` not the upstream `subFlow`.

Reply:
@He-Pin Is it fine to resolve this conversation? I think the issue you point out was specific to https://github.com/apache/incubator-pekko/issues/980 and not about this feature in general?

Reply:
@He-Pin Can you look at the above comment, I want to start working on this today

Reply:
I will review it after work, but I think you can added a `Sink.orElse`, alias with `Flow.orElse` too. Ping me after You think this ready.

And I think your current implementation is wrong.

Reply:
> And I think your current implementation is wrong.

If you can let me know whats wrong about it after work and then I can start writing tests/finishing it off tomorrow

Reply:
I mean I can't see how this will solve the origin problem. when restarting the flow, the connected sink is not restarted.

Reply:
> I mean I can't see how this will solve the origin problem. when restarting the flow, the connected sink is not restarted.

Thats actually the point I am making which is that this PR's is no longer about solving the original problem (i.e. https://github.com/apache/incubator-pekko/issues/980) but is now just generally about supporting `Supervision.restart` for `SubFlow` in way that makes sense. https://github.com/apache/incubator-pekko/issues/980 is going to be solved another way that makes more sense (which we already discussed).

So with that being said, if we had a design goal of adding support for `Supervision.restart` generally in `SubFlow`, does this current implementation make sense or am I missing something?

Reply:
OK, Now I see your point, I will review your PR against this key result.